### PR TITLE
pubsub: add Topic.MessageAs and Subscription.MessageAs

### LINKED ISSUE #657: pubsub/driver: add As functions for Topic, Subscription, and Message
This change was suggested by @vangent on pull #643:

> See blob.Driver for examples.
> 
> Message is used for both reading and writing, which makes it interesting for As. An As function As(i interface{}) bool makes sense for reading, but for writing you want to give access to the underlying type of whatever message is going to be sent to the server. In blob, we do this via WriterOptions.BeforeWrite. So you could:
> 
> Add a BeforeWrite-like function to Message, that's only used when writing
> Split Message into separate types for reading (with As) and writing (with BeforeWrite)
> Add an Options struct with BeforeWrite to SendBatch.

### PR DESCRIPTION
**Pull request by ijt**
Fixes #657

### CODE REVIEW

File: internal/pubsub/driver/driver.go [hunk 1]
```diff
@@ -84,6 +84,13 @@ type Topic interface {
 	// https://github.com/google/go-cloud/blob/master/internal/docs/design.md#as
 	// for more background.
 	As(i interface{}) bool
+
+	// As allows providers to expose provider-specific types.
+	//
+	// See
+	// https://github.com/google/go-cloud/blob/master/internal/docs/design.md#as
+	// for more background.
+	MessageAs(m *Message, i interface{}) bool
```
Comment (jba) [inline internal/pubsub/driver/driver.go]: So I don't think this is right, as discussed on #657.

File: internal/pubsub/gcppubsub/gcppubsub.go [hunk 2]
```diff
@@ -160,3 +171,15 @@ func (s *subscription) As(i interface{}) bool {
 	*c = s.client
 	return true
 }
+
+// MessageAs implements driver.Subscription.MessageAs.
+func (s *subscription) MessageAs(m *driver.Message, i interface{}) bool {
```
Comment (jba) [inline internal/pubsub/gcppubsub/gcppubsub.go]: I think converting back and forth is ultimately going to be lossy and unsatisfying.

If we don't make `driver.Message` an interface, then at least it should have a field
```
Raw interface{}
```
that drivers can populate with the raw value.

File: internal/pubsub/pubsub.go [hunk 3]
```diff
@@ -237,6 +249,14 @@ func (s *Subscription) As(i interface{}) bool {
 	return s.driver.As(i)
 }
 
+// MessageAs converts a message to an implementation-specifc message
+// type, returning true or false depending on whether the conversion
+// succeeded.
+func (s *Subscription) MessageAs(m *Message, i interface{}) bool {
+	dm := toDriverMessage(m)
```
Comment (jba) [inline internal/pubsub/pubsub.go]: Hmm, at this point we've already lost information... not sure how to do this. I guess `pubsub.Message` should hold a `driver.Message` which should hold the raw driver message type.

File: internal/pubsub/gcppubsub/gcppubsub.go (hunk 2 above)
Comment (vangent) [reply internal/pubsub/gcppubsub/gcppubsub.go]: That is not sufficient, as one of the benefits of `As` is that the provider can support multiple raw types. It needs to embed a field here that's an `asFunc`; see `blob.Attributes` for an example.

**Comment by ijt** [comment]
I think this PR is a dead-end. Closing it.

**Pull request closed without merging**